### DISCUSSION TITLE
typescript issue when using useInfiniteQuery with select

### DISCUSSION THREAD

**Discussion by iamhaaamed** [TanStack/query#4941] [Q&A]
I want to konw how can I use useInfiniteQuery with select. This is what I did:
```
export interface User {
  id: number;
  first_name: string;
  last_name: string;
  username: string;
  email: string;
  password?: string;
  branch: string;
  role: 'administrator' | 'user';
  phone?: string;
  avatar?: any;
  active?: number;
}

type Params = { pageParam?: number };
type Response = { users: User[] };

async function getUsers({ pageParam = 1 }: Params): Promise<Response> {
  const response = await client({
    url: `system/users?page=${pageParam}`,
    method: 'GET',
  });
  return response.data;
}

export function useUsers(config: UseInfiniteQueryOptions<Response>) {
  const queryKey = getQueryKey('users');
  return useInfiniteQuery(queryKey, getUsers, {
    getNextPageParam: (lastPage, allPages) => {
      const morePagesExist = lastPage?.users?.length === 10;
      if (!morePagesExist) return false;
      return allPages.length + 1;
    },
    select: (data) => ({
      ...data,
      pages: data.pages.flatMap((page) => page.users),
    }),
    ...config,
  });
}
```

 but it shows me this error:
```
Type '(data: InfiniteData<Response>) => { pages: User[]; pageParams: unknown[]; }' is not assignable to type '(data: InfiniteData<Response>) => InfiniteData<Response>'.
  Call signature return types '{ pages: User[]; pageParams: unknown[]; }' and 'InfiniteData<Response>' are incompatible.
    The types of 'pages' are incompatible between these types.
      Type 'User[]' is not assignable to type 'Response[]'.
        Property 'users' is missing in type 'User' but required in type 'Response'.ts(2322)
```

**Comment by TkDodo** [comment]
show a typescript playground please

**Reply by iamhaaamed** [reply]
> show a typescript playground please

This is the playground:
[https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnm...](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAqgZxQSQDsAzYQ4GFARQFcUpEB5MGYCQnOAXzmKghBwA5AAEYAQw4SAxgGsA9FBTjpMALQBHOgyEBuALAAoUJFgY4NPEVLlKteom69+g0RKkqFSleq0O9RibQ8OIAHmycfALCYREBhkYooabwZJRQxCpoAAriAOZk4nbaiAA8ACoAfBhGcHBKODQANjA4AFxw5QDaALoGhnXSEDSEMB2ENCAARvT9dYRJY3A4MFBkeXAAPnATTU1zcGBKAG5slh0ra4Qb27v7RlxGRkiocADSKIi5UOIgnAC8NQGcC6sk+F1W6x6HUkiH6XH6SRScCGUjgABEAKIAMQAglgADLlAD6BPwAFl8OU4ICAIwABn6iWSwV4I1UbEIcDyKBg9gYHzK1MWKEIABNOILvr8cJUABRgxAQq55AA0h3EPz+AH4OuUAJRAupKGA0KBc0GfdUAOltcrAmplcG1IIdWpwPTgHV6+r6DyehiRrNRKxRTWAovggNiEBw1uk3kocvQtTgU3EeCwACUCR0hLSAEwAdmt9NL1tpbQAHPTa0JVQ99UyEoGWWY0vRMtI0Lh6Ea4MAxeNJjMoAdSFAVsTCL8UMr1gcmhmYNPZ-PrgdLPQZyA58tIRvUygQOJgE113kDg6cDgAO7QMW6-cqg5TH6EaQACwvB34TT3QjiGKIBkMAlxFNAQhbMIW5QPEdRgJ+7AoE+lwLqm4jHEUmpPrCBw+MAxwocO0yzP6LYvDkjp-DS5gOjy0ogE+EykVA3D9JRcBZigOCQBwaCApgsHtNgeBQL07EBhmiAfmyH6sOw3K8r2k7JhqDHUbRtLcB0jE4PqukuGBKClNxvHsHg1QpsCIbwA0fF4LR4i3qe8DSOGkbJqmdRmuecAAAY4IgKzHvIwnavRKD-AAJOgkWMVw-kNsCdS7jASFDsIADimLlPWqZcE2qYmmaXL2RZKDWmK2HwgGQZmMQ7IKVyW4qTgcqoqQeQdL21igcUDjMM1OCmTxDkoJUhrWYMFnwH4AqfLRPJ8iUgpykIwlCEVwIleaFhWCQ-XUCUcrzYggrqstbXqtNdRKTAAByiy5BpWodHKS4rC9KDquIezffpNJWd5d22XAIDQCgAOYuEoaAp9MDfdq1rhda-7XOlNL-HSjIg3UwDEHAcoAIQQ0o0OwzAhq7VymRNHgBx3fUvKlXAf1NADaOinkmMANRwLSjPcMlTN4P+qjvdVEiGv81ReSld22lV2Ei0z6k8R0UviNakVxsQS4wOS4hgHK9r5CgMvVJFKNifpqt1IV9twErnXAGqBXbVwQA)

**Reply by TkDodo** [reply]
`UseInfiniteQueryOptions` takes generics, which you've just left out:

[fixed playground](https://www.typescriptlang.org/play?target=6#code/JYWwDg9gTgLgBDAnmApnA3nAqgZxQSQDsAzYQ4GFARQFcUpEB5MGYCQnAGjlvsQGkUiOAF84xKBBBwA5AAEYAQw5KAxgGsA9FBSLVMALQBHOgxkBuALAAoUJFgY4NPEVLlKvBqPGTp8pSp6Wjp6hiZ8FjZ20PCKAB5sOD5SsvGJkdY2KHH28GSUUMR6aAAKigDmZIoepogAPAAqAHwYNnBwOjg0ADYwOABccA0A2gC6VtbtqhA0hDCDhDQgAEb0E+2E2fNwODBQZOVwAD5wi93d63BgOgBubM6Du-uEhydnFzYiNjZIqHCCiDKUEUICSAF5WpM4MN1EJHnsDqNBspEBMRBNsrk4NMVHAACIAUQAYgBBLAAGQaAH1yfgALL4BpwCEARgADBMsjkYuJZvo2IQ4OUUDBPAIhI04FsUIQACZJAFAkE4JoAClhiHhz3K3DAimBoIA-IMGgBKSHtHQwGhQQUwoTcAB0ztVeoNSUN0LdytGcEGY1N40+32smJ5ON22O6wBl8AhaQgOEdqhClFV6DacGWijwWAASuTBjIWQAmADsjrZlcdLP6AA42Y2ZJxPqbOZlQ9yHPl6EVVGhcPQLXBgLKFktVlBLqQoLsqYQQSgtQdLt0czB54vly9Ls56AuQEudgid5mUCBFMButvypc9TgcAB3aCy43H7WXZbAwiqAAWN8uSRuiPGRFFlEAyGAJ5qmgGRjlkPcoAydowF-dgUDfJ4V0zRQbmqfU3xRS5QmAG4MPHFY1mDDtflKfVlWZRw9WFJUQDfRZKKgUQJlouA8xQHBIA4NAIUwRCBmwPAoDGbiQxzRAf15H9WHYIURUHWd0yuCoUFYxiWVEQZWJwU0jN8KCUDqfjBPYPAWgzKEI3gTohLwRjFEfS94FUaNY3TTN2hta84AAAxwRBdnPTRxMNZiUDBAASdA4tYkQQpbKF2kPGA0LHWQAHECQaZtMxENtMytG1BRc2yUEdWV8LREMwwcYg+RUwU9w0nBVRxUhykGQdXEgmo+GYDqcCsgTXJQbhZnUQgIEfQhuA07hrJm7gSSgYF6iwl4jkVejQSaJpzQcqZbPgcIGABRjhVFWoAVVGRxJkcqoUq20nBcEgRuoWpVRu8VEG4B7uu4C72jUmAADktjKFjjsGVU112RHZrgRRzgxkzmXsgLoacuAQGgXSdJwAkEkjCE0ZgDHDUdGLHWAl4cuZMFWQ5Qn2mAYg4FVABCUmdFxqmoJgc0vsFIpujwS5oY6EUqqxnGKZZmVynZgBqOAWQV0QMsVvBgP0FGGqUc0wRafzMuh516vwo3Fe04UJItxRHTipNiDXGA6UUMBVVdHSrZaOKmakkznfaMqY7gB2+uAHVSo+kQgA)

the problem is though that you are creating a very wide interface. TypeScript would be fine with passing in a new `queryKey` or `queryFn` via options. It's usually not a good abstraction to allow "all options of useQuery" to be passed to your custom hook. There are also options that likely _should_ be the same per observer, like `staleTime`, that you do not want consumers to pass in.

**Reply by daisymaclennan** [reply]
@TkDodo I am struggling to get useInfiniteQuery working due to "Expected 1-2 arguments, but got 3." which is also demonstrated in this fixed playground example you've posted. Is there any way to get around this issue?

**Reply by TkDodo** [reply]
because v5 has a different syntax and this playground is half a year old....

[v5 syntax playground](https://www.typescriptlang.org/play?target=6#code/JYWwDg9gTgLgBDAnmApnA3nAqgZxQSQDsAzYQ4GFARQFcUpEB5MGYCQnAGjlvsQGkUiOAF84xKBBBwA5AAEYAQw5KAxgGsA9FBSLVMALQBHOgxkBuALAAoUJFgY4NPEVLlKvBqPGTp8pSp6Wjp6hiZ8FjZ20PCKAB5sOD5SsvGJkdY2KHH28GSUUMR6aAAKigDmZIoepogAPAAqAHwYNnBwOjg0ADYwOABccA0A2gC6VtbtqhA0hDCDhDQgAEb0E+2E2fNwODBQZOVwAD5wi93d63BgOgBubM6Du-uEhydnFzYiNjZIqHCCiDKUEUICSAF5WpM4MN1EJHnsDqNBspEBMRBNsrk4NMVHAACIAUQAYgBBLAAGQaAH1yfgALL4BpwCEARgADBMsjkYuJZvo2IQ4OUUDBPAIhI04FsUIQACZJAFAkE4JoAClhiHhz3K3DAimBoIA-IMGgBKSHtHQwGhQQUwoTcAB0ztVeoNSUN0LdytGcEGY1N40+32smJ5ON22O6wBl8AhaQgOEdqhClFV6DacGWijwWAASuTBjIWQAmADsjrZlcdLP6AA42Y2ZJxPqbOZlQ9yHPl6EVVGhcPQLXBgLKFktVlBLqQoLsqYQQSgtQdLt0czB54vly9Ls56AuQEudgid5mUCBFMButvypc9TgcAB3aCy43H7WXZbAwiqAAWN8uSRuiPGRFFlEAyGAJ5qmgGRjlkPcoAydowF-dgUDfJ4V0zRQbmqfU3xRS5QmAG4MPHFY1mDDtflKfVlWZRw9WFJUQAoydRAmWi4DzFAcEgDg0AhTBEIGbA8CgMZOJDHNEB-Xkf1YdghRFQdZ3TK4KhQVjREGVicFNPTfCglA6l4-j2DwFoMyhCN4E6AS8EYxRH0veBVGjWN00zdobWvOAAAMcEQXZz00UTDWYlAwQAEnQKLWJEAKWyhdpDxgNCx1kABxAkGmbTMRDbTMrRtQUHMslBHVlfC0RDMMHGIPklMFPc1JwVUcVIcpBkHVxIJqPhmBanAzL4xyUG4WZ1EICBH0Ibg1O4cyJu4EkoGBeosJeeDFXo0Emm4RZKKgJpzRsqZLPgcIGABRjhVFWoAVVGRRJkYqoVK20nBcEgBuoWpvNSuBnWTdhupS9p2hu8VEEhqGYaJQhBge9r4faAbgEUboyhY-bBhZdGVJgAA5LZce0-G4FVNddgp7hsZxrSDOZayfKhuy4BAaBKeFHACQSSMIVpmAKcNR0IsdYCXgy5kwVZDl2Yx4hqYAQm5nQKf5wWYHNL7WrlFA3BQWVLihjoRTKuBGa1qWZXKWWAGo4BZM3RCJvBgP0QZVRqpRzTBFogfN9pQb9xQiZQ5nBnDx0oqTYg1xgOlFDAVVXS0gOWiiiWJIMomivhoq0SAA)

**Comment by EdgarJMesquita** [comment]
This isn't a dealbreaker but indeed annoying, I ended up using something like this

```ts
type CustomInfiniteQueryResult<T> = Omit<UseInfiniteQueryResult<T[], Error>, 'data'> & {
  data: T[];
};


select(data): any { ... }


useInfiniteQuery({
...
select(data): any { ... }
}) as unknown as CustomInfiniteQueryResult<YourType>
```
